modules/images/image-loading-optimization/storage/data.php

```diff
@@ -0,0 +1,125 @@
+<?php
+/**
+ * Metrics storage data.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Gets the expiration age for a given page metric.
+ *
+ * When a page metric expires it is eligible to be replaced by a newer one.
+ *
+ * TODO: However, we keep viewport-specific page metrics regardless of TTL.
+ *
+ * @return int Expiration age in seconds.
+ */
+function ilo_get_page_metric_ttl() {
+	/**
+	 * Filters the expiration age for a given page metric.
+	 *
+	 * @param int $ttl TTL.
+	 */
+	return (int) apply_filters( 'ilo_page_metric_ttl', MONTH_IN_SECONDS );
+}
+
+/**
+ * Unshift a new page metric onto an array of page metrics.
+ *
+ * @param array $page_metrics          Page metrics.
+ * @param array $validated_page_metric Validated page metric.
+ * @return array Updated page metrics.
+ */
+function ilo_unshift_page_metrics( $page_metrics, $validated_page_metric ) {
+	array_unshift( $page_metrics, $validated_page_metric );
+	$breakpoints          = ilo_get_breakpoint_max_widths();
+	$sample_size          = ilo_get_page_metrics_breakpoint_sample_size();
+	$grouped_page_metrics = ilo_group_page_metrics_by_breakpoint( $page_metrics, $breakpoints );
+
+	foreach ( $grouped_page_metrics as &$breakpoint_page_metrics ) {
+		if ( count( $breakpoint_page_metrics ) > $sample_size ) {
+			$breakpoint_page_metrics = array_slice( $breakpoint_page_metrics, 0, $sample_size );
+		}
+	}
+
+	return array_merge( ...$grouped_page_metrics );
+}
+
+/**
+ * Gets the breakpoint max widths to group page metrics for various viewports.
+ *
+ * Each max with represents the maximum width (inclusive) for a given breakpoint. So if there is one number, 480, then
+ * this means there will be two viewport groupings, one for 0<=480, and another >480. If instead there were three
+ * provided breakpoints (320, 480, 576) then this means there will be four viewport groupings:
+ *
+ *  1. 0-320 (small smartphone)
+ *  2. 321-480 (normal smartphone)
+ *  3. 481-576 (phablets)
+ *  4. >576 (desktop)
+ *
+ * @return int[] Breakpoint max widths, sorted in ascending order.
+ */
+function ilo_get_breakpoint_max_widths() {
+
+	/**
+	 * Filters the breakpoint max widths to group page metrics for various viewports.
+	 *
+	 * @param int[] $breakpoint_max_widths Max widths for viewport breakpoints.
+	 */
+	$breakpoint_max_widths = array_map(
+		static function ( $breakpoint_max_width ) {
+			return (int) $breakpoint_max_width;
+		},
+		(array) apply_filters( 'ilo_breakpoint_max_widths', array( 480 ) )
```

Comment:
WordPress's own responsive preview in the block editor is actually even smaller at 360 for mobile. And 480 is wider than all of the widths for mobile devices in Chrome DevTools. I picked 480 from doing various searches, though I can't recall the specific resource I used. Some references for 480 pixels:

* https://www.lambdatest.com/blog/how-to-use-css-breakpoints-for-responsive-design/#:~:text=What%20are%20common%20breakpoints?,extra%20large%20screens%20like%20TV.
* https://testsigma.com/blog/css-breakpoints/#:~:text=While%20there%20is%20no%20hard%20and%20fast%20rule%20for%20what%20breakpoints%20should%20be%20used%2C%20values%20of%20common%20CSS%20breakpoints%20are%20320px%20or%20480px%20(for%20mobile%20phones)%2C%20768px%20(for%20tablets)%2C%20and%201920px%20(for%20desktop%20computers).%20Ultimately%2C%20it%20is%20up%20to%20the%20designer%20to%20decide%20what%20breakpoints%20make%20sense%20for%20their%20project.
* https://testsigma.com/blog/css-breakpoints/#:~:text=from%2C%20or%20to.-,Which%20CSS%20breakpoints%20to%20use%3F,-Although%20every%20device

Reply:
i was going to ask the same question as I am sure others will so it would be good to eventually have a concise answer for why the specific values were chosen (and in what instances you might want to alter them)

Reply:
Thanks both, for now 480 is good IMO. We can always adjust later, and there's a filter anyway.

Reply:
I discovered that this number is indeed too small, at least on its own. The columns block, for example, changes its layout at the 786px breakpoint. Gutenberg includes a [few breakpoints defined](https://github.com/WordPress/gutenberg/blob/093d52cbfd3e2c140843d3fb91ad3d03330320a5/packages/base-styles/_breakpoints.scss#L11-L13), including for mobile (480px), small (600px), and medium (786px). These seem to be used the most for media queries on the frontend, so I'm now using them in https://github.com/WordPress/performance/pull/884